unified_planning/model/action.py

```diff
@@ -402,16 +419,28 @@ def _add_effect_instance(self, effect: "up.model.effect.Effect"):
         assert (
             effect.environment == self._environment
         ), "effect does not have the same environment of the action"
-        if not effect.is_conditional():
+        if not effect.is_conditional() and not effect.value.type.is_bool_type():
             if effect.is_assignment():
-                if (
-                    effect.fluent in self._fluents_assigned
-                    or effect.fluent in self._fluents_inc_dec
-                ):
+                assigned_value = self._fluents_assigned.get(effect.fluent, None)
+                # if the same fluent is involved in an increase/decrease, raise exception
+                if effect.fluent in self._fluents_inc_dec:
                     raise UPConflictingEffectsException(
-                        f"The effect {effect} is in conflict with the effects already in the action."
+                        f"The effect {effect} is in conflict with the increase/decrease effects already in the action."
                     )
-                self._fluents_assigned.add(effect.fluent)
+                # the same fluent is involved in another assign
+                elif assigned_value is not None:
+                    # if the 2 values are different, raise exception
+                    if assigned_value != effect.value and not (
+                        assigned_value.is_constant()
+                        and effect.value.is_constant()
+                        and assigned_value.constant_value()
+                        == effect.value.constant_value()
+                    ):
+                        raise UPConflictingEffectsException(
+                            f"The effect {effect} is in conflict with the effects already in the action."
+                        )
+                else:
+                    self._fluents_assigned[effect.fluent] = effect.value
```

Comment:
@Framba-Luca This piece of code is duplicated in 3 different files. Can we refactor it?